### PR TITLE
included soil ODE scheme. Little issue in GW mass balance

### PR DESCRIPTION
**Pull request by RY4GIT**
## Summary
This PR addresses the issue https://github.com/NOAA-OWP/cfe/issues/82

Note that this PR will solve the issue above, but changes the model behavior and scheme largely, and increase the computational time. 

## Known issue 
This PR still has a little issue in groundwater mass balance in this PR. Refer to the code https://github.com/RY4GIT/SMSigxModel/tree/mcmillan_ws/py_cfe or https://github.com/NOAA-OWP/cfe/pull/66 or , which don't have groundwater mass balance issues. I probably forgot to copy or delete some lines from those versions to this one.

### CODE REVIEW

File: bmi_cfe.py [hunk 1]
```diff
@@ -368,14 +379,14 @@ def finalize_mass_balance(self, verbose=True):
 
         self.vol_soil_end = self.soil_reservoir['storage_m']
         
-        self.global_residual  = self.volstart + self.volin - self.volout - self.volend -self.vol_end_giuh
-        self.schaake_residual = self.volin - self.vol_sch_runoff - self.vol_sch_infilt - self.vol_et_from_rain
+        self.global_residual  = self.volstart + self.volin - self.volout - self.volend -self.vol_end_giuh #(?) - vol_in_nash_end? --- This is already excluded from the system.
```
Comment (jmframe) [inline bmi_cfe.py:382]: I'm not 100% sure, but I think this is still needed here. Since the global residual needs to account for anything that can be left over. Unless you are sure that it will be zero at any time step, we should still subtract from volstart and volin.

File: bmi_cfe.py [hunk 2]
```diff
@@ -233,7 +240,7 @@ def initialize(self,current_time_step=0):
         
         # ________________________________________________
         # Nash cascade        
-        self.K_nash = 0.03
+        self.K_nash = self.K_nash
```
Comment (jmframe) [inline bmi_cfe.py:243]: Can just delete.

File: cfe.py [hunk 3]
```diff
@@ -3,6 +3,65 @@
 import pandas as pd
 import sys
 
+from scipy.integrate import odeint
+from numba import jit
+import math
+
+@jit(nopython=True)
+def conceptual_reservoir_flux_calc(t, S, storage_threshold_primary_m, storage_max_m, coeff_primary, coeff_secondary, PET, infilt, wltsmc):
```
Comment (jmframe) [inline cfe.py:11]: Why not include this in the CFE class? Then all these inputs will be in CFE_STATE.

File: cfe.py [hunk 4]
```diff
@@ -24,35 +83,26 @@ def run_cfe(self, cfe_state):
             self.et_from_rainfall(cfe_state)
         
         cfe_state.vol_et_from_rain += cfe_state.actual_et_from_rain_m_per_timestep
-        #cfe_state.vol_et_to_atm += cfe_state.actual_et_from_rain_m_per_timestep
+        cfe_state.vol_et_to_atm += cfe_state.actual_et_from_rain_m_per_timestep
         cfe_state.volout += cfe_state.actual_et_from_rain_m_per_timestep
 
-        # ________________________________________________
-        # Calculate evaporation from soil
```
Comment (jmframe) [inline cfe.py:31]: We need to include options for all the old code. We have to be able to run an identical version as the C code. So there needs to be an option from the configuration file to run this.

**Review by jmframe — CHANGES_REQUESTED** [review]
Thanks, this is great. I ran the code and it works. The main thing is that this functionality has to be an option, but replacing the old code. Even though the old code has problems, there is an important aspect of Nextgen to be able to run models as the origional author intended, even if it is a mistake, but we'll just have the option to run the soil-ode instead, and we'll run that most times.

So basically we need a setup like the partitioning scheme can be either Xinanjiang or Schaake. We'll need a soil reservoir flux option for "ODE". I can help with this, if you want.

File: bmi_cfe.py (hunk 2 above)
Comment (RY4GIT) [reply bmi_cfe.py:243]: Agree

**Comment by RY4GIT** [comment]
Thank you so much for the review! 

If you could help me with separating ODE option, it would be great! And I can go ahead figuring out the GW mass balance issue. 

Do all changes need to be solved and the PR needs to be merged before you work on the ODE option? Then I will prioritize it (probably done by EOD today or tomorrow)

**Comment by jmframe** [comment]
I can make a new branch and implement the option in the configuration file.
Then I'll share that with you and you can decide implement if you want or
re-do it.

**Comment by RY4GIT** [comment]
Sounds good!
sent from mobile

File: cfe.py (hunk 3 above)
Comment (RY4GIT) [reply cfe.py:11]: I simply didn't try! They can be included in the CFE class
